source-batchsql/main.go

```diff
@@ -0,0 +1,480 @@
+			ResourceConfigJson: resourceConfigJSON,
+			DocumentSchemaJson: minimalSchema,
+			Key:                defaultKey,
+		})
+	}
+
+	return &pc.Response_Discovered{Bindings: bindings}, nil
+}
+
+var catalogNameSanitizerRe = regexp.MustCompile(`(?i)[^a-z0-9\-_.]`)
+
+func recommendedCatalogName(schema, table string) string {
+	var catalogName string
+	// Omit 'default schema' names for Postgres and SQL Server. There is
+	// no default schema for MySQL databases.
+	if schema == "public" || schema == "dbo" {
+		catalogName = table
+	} else {
+		catalogName = schema + "_" + table
+	}
+	return catalogNameSanitizerRe.ReplaceAllString(strings.ToLower(catalogName), "_")
+}
+
+func (driver) Validate(ctx context.Context, req *pc.Request_Validate) (*pc.Response_Validated, error) {
+	var cfg Config
+	if err := pf.UnmarshalStrict(req.ConfigJson, &cfg); err != nil {
+		return nil, fmt.Errorf("parsing endpoint config: %w", err)
+	}
+
+	// Validate connection to database
+	log.WithFields(log.Fields{
+		"address":  cfg.Address,
+		"user":     cfg.User,
+		"database": cfg.Database,
+	}).Info("connecting to database")
+	var db, err = sql.Open("pgx", cfg.ToURI())
+	if err != nil {
+		return nil, fmt.Errorf("error opening database connection: %w", err)
+	} else if err := db.PingContext(ctx); err != nil {
+		return nil, fmt.Errorf("error pinging database: %w", err)
+	}
+	defer db.Close()
+
+	var out []*pc.Response_Validated_Binding
+	for _, binding := range req.Bindings {
+		var res Resource
+		if err := pf.UnmarshalStrict(binding.ResourceConfigJson, &res); err != nil {
+			return nil, fmt.Errorf("parsing resource config: %w", err)
+		}
+
+		out = append(out, &pc.Response_Validated_Binding{
+			ResourcePath: []string{res.Name},
+		})
+	}
+	return &pc.Response_Validated{Bindings: out}, nil
+}
+
+func (driver) Pull(open *pc.Request_Open, stream *boilerplate.PullOutput) error {
+	var cfg Config
+	if err := pf.UnmarshalStrict(open.Capture.ConfigJson, &cfg); err != nil {
+		return fmt.Errorf("parsing endpoint config: %w", err)
+	}
+
+	var resources []Resource
+	for _, binding := range open.Capture.Bindings {
+		var res Resource
+		if err := pf.UnmarshalStrict(binding.ResourceConfigJson, &res); err != nil {
+			return fmt.Errorf("parsing resource config: %w", err)
+		}
+		resources = append(resources, res)
+	}
+
+	if open.StateJson != nil {
+		// TODO: Consider implementing persistent state, based on the resource name property.
+	}
+
+	var capture = &capture{
+		Config:    cfg,
+		Resources: resources,
+		Output:    stream,
+	}
+	return capture.Run(stream.Context())
+}
+
+type capture struct {
+	Config    Config
+	Resources []Resource
+	Output    *boilerplate.PullOutput
+}
+
+func (c *capture) Run(ctx context.Context) error {
+	log.WithFields(log.Fields{
+		"address":  c.Config.Address,
+		"user":     c.Config.User,
+		"database": c.Config.Database,
+	}).Info("connecting to database")
+	var db, err = sql.Open("pgx", c.Config.ToURI())
```

Comment:
The `sql.Open` call and associated config-to-DSN mapping is essentially the only part of this connector which is database-specific (well, also the datatype mapping stuff potentially but I haven't investigated that yet), so in principle almost all of this logic could be factored out into a helper package and the per-database connector code would be 100-200 lines saying essentially "here's what a configuration looks like, here's a function that takes a config and returns a generic `*sql.DB`, and here's how to translate output values".